lib/runtime/adapter-read.js

```diff
@@ -114,6 +115,27 @@ class AdapterRead {
         return errors.runtimeError(code, info);
     }
 
+    // Helper to obtain a throttled function that warns on late arrivals.
+    throttledWarnLate(interval) {
```

Comment:
These seem a bit overly-specific, how about just a method `throttledWarn(errorCode, info)` that adapters could call with whatever error code and info they want?